stdlib/std.ncl

```diff
@@ -1103,6 +1130,254 @@
+              ```
+            "%
+          = fun Domain Codomain label function =>
+            fun arg =>
+              let arg_with_ctr = std.contract.apply Domain label arg in
+              function arg_with_ctr
+              |> contract.apply (Codomain arg_with_ctr) label,
+
+        PosNumber
+          | doc m%"
+              **Warning**: this is an unstable item. It might be renamed,
+              modified or deleted in any subsequent minor Nickel version.
+
+              A contract for a positive number.
+
+              This contract blames only if the value is a number but is not
+              positive: the fact that the value is a number is assumed to be
+              checked by an associated static type annotation.
+            "%
+          = fun label value =>
+            if %typeof% value == 'Number then
+              if value >= 0 then
+                value
+              else
+                %blame% (%label_append_note% "Expected a positive number, got %{%to_str% value}" label)
+            else
+              value,
+
+        RangeStep
+          | doc m%"
+              **Warning**: this is an unstable item. It might be renamed,
+              modified or deleted in any subsequent minor Nickel version.
+
+              A contract for a range step, which must be a positive number.
+              Mostly a wrapper around `std.contract.unstable.PosNumber` with a
+              custom error message.
+
+              This contract blames only if the value is a number but is not
+              positive: the fact that the value is a number is assumed to be
+              checked by an associated static type annotation.
+            "%
+          = fun label =>
+            PosNumber (%label_with_message% "invalid range step" label),
+
+        RangeFun
+          | doc m%"
+              **Warning**: this is an unstable item. It might be renamed,
+              modified or deleted in any subsequent minor Nickel version.
+
+              A function contract which checks that the two first argument are a
+              valid range, that is two numbers `start` and `end` such that
+              `start <= end`.
+
+              This contract blames only if the arguments are of the right type
+              but don't satisfy the tested condition. The type of arguments is
+              assumed to be checked by an associated static type annotation. For
+              example, if the first argument is a string, this contract silently
+              passes.
+            "%
+          =
+            # capture the `contract.label` module to avoid name collisions with
+            # local contract argument `label`
+            let label_module = label in
+            let attach_message = label_module.with_message "invalid range" in
+
+            fun Codomain =>
+              let RangeSecond = fun start label value =>
+                if %typeof% start == 'Number && %typeof% value == 'Number then
```

Comment:
Do we need this check if we're relying on an additional static type annotation anyway?

Reply:
Yes, because of the way function contracts work, the order of checks for the arguments are actually reversed:

```nickel
foo
  | Domain1 -> Codomain1
  | Domain2 -> Codomain2
```

which will be elaborated as (more or less)

```
let first_app = fun x =>
  Codomain1 (f (Domain1 x))
in
first_app | Domain2 -> Codomain2

===>

fun x =>
  Codomain2 (first_app (Domain2 x))
```

So, the contracts are applied in that order, because the second function contract is the outer one:

- `Domain2` to `x`
- `Domain1` to `x`
- `Codomain1` to `f x`
- `Codomain2` to `f x`

So, when the new additional contracts are being applied, the static type annotation hasn't kicked in yet. This was observed in practice as well during testing. The additional contract thus have to yield the control flow to the other one, without failing, and without performing any operation that could cause another primop error (like trying to compare a number and a string).

Reply:
Oh, I see. Thanks for explaining. This might be subtle enough to deserve an explanation somewhere canonical, maybe the manual.

Reply:
Do you mean about applying several function contract in general, or about the "differential approach" specifically?

Reply:
It's about the evaluation order of several function contracts applied to a field. That has implications about this differential approach, namely what can be relied on by which contract, but I feel it's a more general point to make.

Reply:
Ok. I'll proceed with this PR but open a corresponding issue.